File: src/cc/Common/Status.cc
/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for Status.
/// This file contains declarations for Status, a class to hold Nagios-style
/// status information for a running program.

#include <Common/Compat.h>

#include "Status.h"

#include <Common/Serialization.h>

using namespace Hypertable;
using namespace std;

const char *Status::code_to_string(Code code) {
  switch (code) {
  case Code::OK:
    return "OK";
  case Code::WARNING:
    return "WARNING";
  case Code::CRITICAL:
    return "CRITICAL";
  default:
    break;
  }
  return "UNKNOWN";
}

uint8_t Status::encoding_version() const {
  return 1;
}

size_t Status::encoded_length_internal() const {
  return 4 + Serialization::encoded_length_vstr(m_text);;
}

/// Serialized format is as follows:
/// <table>
///   <tr><th>Encoding</th><th>Description</th></tr>
///   <tr><td>i32</td>%Status code</td></tr></tr>
///   <tr><td>vstr</td>%Status text</td></tr>
/// </table>
void Status::encode_internal(uint8_t **bufp) const {
  Serialization::encode_i32(bufp, static_cast<int32_t>(m_code));
  Serialization::encode_vstr(bufp, m_text);
}

void Status::decode_internal(uint8_t version, const uint8_t **bufp,
                             size_t *remainp) {
  lock_guard<mutex> lock(m_mutex);
  m_code = static_cast<Code>(Serialization::decode_i32(bufp, remainp));
  m_text = Serialization::decode_vstr(bufp, remainp);
}

File: java/hypertable-common/src/main/java/org/hypertable/Common/Status.java
/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

package org.hypertable.Common;

import java.io.UnsupportedEncodingException;
import java.net.ProtocolException;
import java.nio.ByteBuffer;
import java.util.zip.DataFormatException;

public class Status extends Serializable {

  /** Enumeration for status codes. */
  public enum Code {
    OK(0), WARNING(1), CRITICAL(2), UNKNOWN(3);
    private final int value;
    private Code(int value) {
      this.value = value;
    }
    public int getValue() {
      return value;
    }
  };

  /// Status text string constants.
  public class Text {
    static public final String SERVER_IS_COMING_UP = "Server is coming up";
    static public final String SERVER_IS_SHUTTING_DOWN = "Server is shutting down";
    static public final String STANDBY = "Standby";
  };

  /** Constructor.
   */
  public Status() { }

  /** Copy constructor.
   * @param other Other status object to make copy from
   */
  public Status(Status other) {
    synchronized (other) {
      mCode = other.mCode;
      mText = other.mText;
    }
  }


  /** Sets status code and text
   * @param code %Status code
   * @param text %Status text
   */
  public synchronized void set(Code code, String text) {
    mCode = code;
    mText = text;
  }

  /** Gets status code.
   * @return Status code
   */
  public synchronized Code get() { return mCode; }

  /** Returns encoding version.
   * @return Encoding version
   */
  protected byte encodingVersion() {
    return 1;
  }

  /** Returns internal serialized length.
   * This function is to be overridden by derived classes and should return
   * the length of the the serialized object per-se.
   * @return Internal serialized length
   * @see encodeInternal() for encoding format
   */
  protected int encodedLengthInternal() throws UnsupportedEncodingException {
    return 4 + Serialization.EncodedLengthVStr(mText);
  }

  /** Writes serialized representation of object to a buffer.
   * This function is to be overridden by derived classes and should encode
   * the object per-se.
   * @param buf ByteBuffer destination
   */
  protected void encodeInternal(ByteBuffer buf) throws UnsupportedEncodingException {
    buf.putInt(mCode.getValue());
    Serialization.EncodeVStr(buf, mText);
  }

  /** Reads serialized representation of object from a buffer.
   * This function is to be overridden by derived classes and should decode
   * the object per-se as encoded with encodeInternal().
   * @param version Encoding version
   * @param buf ByteBuffer source
   * @see encodeInternal() for encoding format
   */
  protected void decodeInternal(byte version, ByteBuffer buf) throws ProtocolException, DataFormatException, UnsupportedEncodingException {
    mCode = Code.values()[buf.getInt()];
    mText = Serialization.DecodeVStr(buf);
  }

  /** Status code */
  private Code mCode = Code.OK;

  /** Status text */
  private String mText;


}
